### ISSUE TITLE
Suggestion: support directories on drag and drop

### ISSUE THREAD

**Issue by michaelmdr** [hackingharold/ngx-dropzone#60]
Hello Harold, I really appreciate your great work!

I would like to implement the ability to extract files from folders and subfolders using drag and drop.

If you don't mind, I'll start a pull request in the near future.

Greetings from Bavaria!

**Comment by hackingharold** [comment]
Hi there,

I just released the latest version for Angular v17 which includes support for dropping folders by default.
Please let me know if this helps you. I used your implementation for inspiration.

[Releases](https://github.com/hackingharold/ngx-dropzone/releases)

**Comment by michaelmdr** [comment]
Hello there,

I just tested it, it works great! Thank you very much.

One detail i noticed is that in Chrome if you drop a folder containing more than 100 files, it just gets 100 files and not the whole content. For that issue i wrote the `_breakDirectoryReaderFilesLimit`* method in my Pull Request to solve this problem. It would be nice to also solve this within your solution.

It is arguably not the prettiest solution, but it does its job (maybe you are able to find a better solution):
` 
private async _breakDirectoryReaderFilesLimit(directoryEntry: FileSystemDirectoryEntry) {
  const reader = directoryEntry.createReader();
  let resultEntries: FileSystemEntry[] = [];

  const recurse = async () => {
    const entries = await this._readDirectoryEntries(reader);
    if ((entries).length > 0) {
      resultEntries = resultEntries.concat(entries);
      await recurse();
    }
  };

  await recurse();
  return resultEntries;
}
`

Thank you in  advance,
I'm looking forward to your solution!

**Comment by hackingharold** [comment]
Hi @michaelmdr,

thanks for your feedback. As stated in [MDN](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryReader/readEntries), the `readEntries()` method will only return the first 100 entries.
I am just curious, what's your use case for uploading more than a hundred files using a single file input?

In my opinion, we should use an optional `limit` property to keep this default behaviour for users which are less involved in this topic.

**Comment by hackingharold** [comment]
Hi @michaelmdr,

I just released v17.2.0 which skips the limit by default.

After thinking about it, I decided not to create a property for this. We're basically creating a consistent cross-browser experience by ignoring Chrome's behavior here.

**Comment by alexandis** [comment]
How to obtain the relative path of each file being uploaded?

**Comment by alexandis** [comment]
Would be great to include file relative path information (now it is always ''). The reason this might be needed it if we need to recreate a directory structure on server.

So I had to use an ugly method for that - in fact, now the original field (`formControl`) is useless:(

```
onDropCases(event: DragEvent) {
    var items = event.dataTransfer.items;
    this.files = [];  
    const traverseDirectory = (entry: FileSystemEntry, path: string): Promise<void> => {
      return new Promise<void>((resolve, reject) => {
        if (entry.isFile) {
          const fileEntry = entry as FileSystemFileEntry;
          fileEntry.file((file: File) => {
            this.files.push(this.copyFile(file, fileEntry.fullPath));
            resolve();
          });
        }
        else if (entry.isDirectory) {
          const directoryEntry = entry as FileSystemDirectoryEntry;
          const dirReader = directoryEntry.createReader();
          dirReader.readEntries(entries => {  
            const promises = entries.map(childEntry => traverseDirectory(childEntry, path + entry.name + '/'));
            Promise.all(promises).then(() => resolve());
          });
        }
      });
    };  
    const promises = [];
    for (var i = 0; i < items.length; i++) {
      var item = items[i];
      if (item.kind === 'file') {
        var webkitGetAsEntry = item.webkitGetAsEntry();
        promises.push(traverseDirectory(webkitGetAsEntry, ''));
      }
    }
    Promise.all(promises).then(() => {});
  }
```

**Comment by alexandis** [comment]
Uhm... too much on my plate now, sorry. But the client wants drag-n-drop mechanism. Now - just a quick question: is there currently some limitation on the amount of files I can drag-n-drop. Just checked a multiselection now using a "file" mode, the folder contained 700+ files. All of them were uploaded (to memory). However, when I drag-n-droped the folder, only 100 were... I wonder what's wrong: is it my method above or something deeper in the component code...

**Comment by alexandis** [comment]
Hi @hackingharold.

Drag-n-drop behavior was a client requirement (as usual it happens :-D).
Besides, I have noticed, that Native OS dialog does not allow selecting multiples folders (which is strange, since `multiple` attribute is present).

All in all, I am stick to my own custom method to handle `drop` event. But I've noticed, that I bumped onto the same limitation which was mentioned before, 100 files per `readEntries`. Does it mean I need to copy-paste the code already implemented in your library?

**Comment by hackingharold** [comment]
Yes, you cannot select multiple directories from the file picker. The `multiple` attribute only works for files, which is a native behavior.

To work around the file limit, I've written the [`readDirectoryWithoutLimit`](https://github.com/hackingharold/ngx-dropzone/blob/a2b6de0fa944694e8e717c1234d17d1d76b827fe/projects/cdk/src/lib/dropzone/dropzone.service.ts#L63) method, maybe this helps.
